frontend/catalyst/jax_tracer.py

```diff
@@ -1077,23 +1078,26 @@ def trace_quantum_measurements(
         if isinstance(output, MeasurementProcess):
 
             # Check if the measurement is supported shot-vector where num_of_total_copies > 1
-            if shots_obj.has_partitioned_shots and not isinstance(
-                output, qml.measurements.SampleMP
-            ):
+            if shots_obj.has_partitioned_shots and not isinstance(output, SampleMP):
                 raise NotImplementedError(
                     f"Measurement {type(output).__name__} does not support shot-vectors. "
                     "Use qml.sample() instead."
                 )
 
             if device.wires is None:
-                d_wires = num_qubits_p.bind()
+                # Automatic qubit management, need to query num qubits for shape-ful measurements
+                # Note: 2 ifs instead of `and` because the next case uses `is_dynamic_wires`, which
+                # asserts that it is not seeing `None`
+                if type(output) in [SampleMP, CountsMP, ProbabilityMP, StateMP]:
+                    d_wires = num_qubits_p.bind()
             elif catalyst.device.qjit_device.is_dynamic_wires(device.wires):
                 d_wires = num_qubits_p.bind()
```

Comment:
I'm slightly hesitant. You're right currently that the variable `d_wires` or rather `nqubits` is only used for those MPs (which is actually most of them, only expval/var is excluded). But I wonder if it isn't more trouble than it's worth if the code below changes (another implicit assumption, might not be obvious why `nqubits` is None in that case). You also didn't apply this to the branch right below which should have the same constraints.

I'd say for simplicity there is no harm in always generating if we can just DCE it.